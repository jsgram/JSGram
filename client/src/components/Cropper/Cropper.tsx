import React from 'react';
import Avatar from 'react-avatar-edit';
import noAvatar from '../assets/noAvatar.svg';
import { Button, Spinner } from 'reactstrap';

interface ILocalState {
    preview: null | string;
    src: null | string;
}

export default class Cropper extends React.Component<any> {

    public FILE_SIZE: number = 2000000;

    public dropZoneStyle: any = {
        width: '100%',
        margin: '0 auto',
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        padding: 20,
        borderWidth: 2,
        borderRadius: 2,
        borderColor: '#eeeeee',
        borderStyle: 'dashed',
        backgroundColor: '#fafafa',
        color: '#bdbdbd',
        outline: 'none',
        transition: 'border .24s ease-in-out',
    };

    public state: ILocalState = {
        preview: null,
        src: null,
    };

    public onClose = (): void => {
        this.setState({preview: null});
    }

    public onCrop = async (preview: any): Promise<void> => {
        this.setState({preview});
        this.props.createFile(preview);
    }

    public onBeforeFileLoad = (elem: any): void => {
        if (elem.target.files[0].size > this.FILE_SIZE) {
            alert('File is too big!');
            elem.target.value = '';
        }
    }

    public onClick = (): void => {
        this.props.uploadPostAvatar(this.props.avatar);
        this.props.modalToggle();
    }

    public render(): JSX.Element {
        return (
            <div className='text-center' style={{maxWidth: 560, margin: '0 auto'}}>
                <div className='text-center'>
                    <img
                        height={100}
                        width={100}
                        className='mb-3'
                        src={this.state.preview || noAvatar}
                        alt='Preview'
                    />
                </div>

                <Avatar
                    width={450}
                    height={300}
                    imageWidth={300}
                    imageHeight={300}
                    borderStyle={this.dropZoneStyle}
                    shadingColor='white'
                    onCrop={this.onCrop}
                    onClose={this.onClose}
                    onBeforeFileLoad={this.onBeforeFileLoad}
                    src={this.state.src || ''}
                />
                {this.props.loading ? (
                    <Spinner className='mt-3' color='dark'/>
                ) : (
                    <Button
                        className='mt-3'
                        outline
                        color='danger'
                        size='lg'
                        onClick={this.onClick}
                    >
                        Save avatar
                    </Button>
                )}
            </div>
        );
    }
}
