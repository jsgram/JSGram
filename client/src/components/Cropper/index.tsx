import React from 'react';
import Cropper from 'react-easy-crop';
import {Container, Row, Spinner} from 'reactstrap';
import AddPostDropZone from '../AddPost/AddPostDropZone';
import '../AddPost/PostPost/style.scss';
import { createBlobUrl, getCroppedImg } from '../../helpers/upload.photo';

export interface IAddAvatarCropperProps {
    croppedImage: string;
    uploadPostAvatar: (croppedImage: string) => void;
    setCroppedImageForAvatar: (croppedImage: string) => void;
    loading: boolean;
    informFileError: any;
    resetAddPost: any;
}

interface IState {
    imageSrc: string;
    crop: { x: number, y: number };
    zoom: number;
    maxZoom: number;
    aspect: number;
    croppedAreaPixels: null;
}

interface IArea {
    width: number;
    height: number;
    x: number;
    y: number;
}

export interface ILocation {
    x: number;
    y: number;
}

export default class AddPostCropper extends React.Component<IAddAvatarCropperProps> {

    public state: IState = {
        imageSrc: '',
        crop: {x: 0, y: 0},
        zoom: 1,
        maxZoom: 2,
        aspect: 3 / 3,
        croppedAreaPixels: null,
    };

    // 1 Select image
    public onUploadImageToCropper = (imageFile: File): void => {
        createBlobUrl(imageFile, (data: any) => {
            this.setState({imageSrc: data.target.result});
        });
    }

    // 2 Set crop position to Cropper
    public onCropChange = (crop: ILocation): void => {
        this.setState({crop});
    }

    // 3 Set zoom to Cropper
    public onZoomChange = (zoom: number): void => {
        this.setState({zoom});
    }

    public resetImageSrc = (): void => {
        this.setState({imageSrc: ''});
    }

    // 4 Save crop area
    public onCropComplete = (croppedArea: IArea, croppedAreaPixels: IArea): void => {
        this.setState({croppedAreaPixels});
    }

    // 5 Create new image (not file) inside getCroppedImg
    // 6 Create cropped image in url inside getCroppedImg
    // 7 Transform to blob url inside getCroppedImg
    // 8 Transform cropped img in base64 to file
    public onShowCroppedImage = async (): Promise<void> => {
        const cropped = await getCroppedImg(
            this.state.imageSrc,
            this.state.croppedAreaPixels,
        );
        await createBlobUrl(cropped, (data: any) => {
            this.props.setCroppedImageForAvatar(data.target.result);
            this.props.uploadPostAvatar(this.props.croppedImage);
        });
    }

    public render(): JSX.Element {
        const {imageSrc, crop, zoom, maxZoom, aspect}: IState = this.state;
        return (
            <div className='text-center'>

                <Container>
                    <Row>
                        <div className=' mx-auto mt-3 post-label'>New post</div>
                    </Row>
                    <div className='cropper-photo mt-3 mx-auto'>
                        {
                            imageSrc ?
                                (
                                    <div>
                                        < Cropper
                                            image={imageSrc}
                                            crop={crop}
                                            zoom={zoom}
                                            maxZoom={maxZoom}
                                            aspect={aspect}
                                            onCropChange={this.onCropChange}
                                            onCropComplete={this.onCropComplete}
                                            onZoomChange={this.onZoomChange}
                                            cropShape='round'
                                        />
                                    </div>
                                ) : (
                                    <AddPostDropZone
                                        uploadImageToCropper={this.onUploadImageToCropper}
                                        informFileError={this.props.informFileError}
                                        resetImageSrc={this.props.resetAddPost}
                                        sizeMB={2}
                                    />
                                )
                        }
                    </div>
                </Container>

                <Row className='justify-content-center post mx-auto'>
                    <button
                        className='mt-3 ml-0 button'
                        onClick={this.onShowCroppedImage}
                        disabled={!imageSrc}
                    >
                        {this.props.loading ? <Spinner color='white'/> : 'Save'}
                    </button>
                </Row>
            </div>
        );
    }
}
