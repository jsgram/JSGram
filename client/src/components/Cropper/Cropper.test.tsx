import Cropper from './index';
import { shallow } from 'enzyme';
import React from 'react';

describe('Cropper component:', () => {
    let renderer: any;

    const props = {
        croppedImage: 'some value',
        description: 'some value',
        loading: true,
        user: {
            posts: 1,
            followers: [],
            following: [],
            description: 'some value',
            fullName: 'some value',
            username: 'some value',
            photo: 'some value',
            email: 'some value',
            _id: 'some value',
        },
        uploadPostAvatar: jest.fn(() => ('some value')),
        setCroppedImageForAvatar: jest.fn(() => ('some value')),
        resetAddPost: jest.fn(() => ('some value')),
        toggleModal: jest.fn(() => ('some value')),
        informFileError: jest.fn(() => ('some value')),
    };

    beforeEach(() => {
        renderer = shallow(<Cropper {...props}/>);
    });

    test('on crop change - success', () => {
        renderer.instance().onCropChange({x: 1, y: 1});
        expect(renderer.state('crop')).toEqual({x: 1, y: 1});
    });

    test('on zoom change - success', () => {
        renderer.instance().onZoomChange(1);
        expect(renderer.state('zoom')).toBe(1);
    });

    test('reset image src - success', () => {
        renderer.instance().resetImageSrc();
        expect(renderer.state('imageSrc')).toBe('');
    });

    test('crop complete - success', () => {
        renderer.instance().onCropComplete({width: 1, height: 1, x: 1, y: 1}, {width: 1, height: 1, x: 1, y: 1});
        expect(renderer.state('croppedAreaPixels')).toEqual({height: 1, width: 1, x: 1, y: 1});
    });

    test('render-success', () => {
        expect(renderer).toMatchSnapshot();
    });
});
