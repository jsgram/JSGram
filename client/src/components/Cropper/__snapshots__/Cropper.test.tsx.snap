// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cropper component: render-success 1`] = `ShallowWrapper {}`;
